Mathlib/Algebra/Category/ModuleCat/Differentials/Presheaf.lean

```diff
@@ -0,0 +1,257 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.Algebra.Category.ModuleCat.Presheaf
+import Mathlib.RingTheory.Kaehler
+
+/-!
+# The presheaf of differentials of a presheaf of modules
+
+In order to do this, we first introduce the presheaf
+of absolute differentials `absoluteDifferentials R` (i.e. the differentials over `ℤ`).
+
+## Main definitions
+
+- Given a presheaf of modules `M` over a presheaf of commutative rings, we define
+the type `M.AbsoluteDerivation` of absolute derivations from `R` to `M`.
+
+## TODO
+
+- If `F : C ⥤ D` is a functor, and `f : S ⟶ F.op ⋙ R` a morphism of presheaves of rings,
+construct the relative differentials of `f` as the quotient of `absoluteDifferentials R`
+by the image of the canonical morphism from the pullback of `absoluteDifferentials S`:
+this requires the construction of the pullback functor for presheaves of modules
+(as a left adjoint to the functor constructed in the file
+`Algebra.Category.ModuleCat.Presheaf.Pushforward`).
+- Sheafify these constructions
+
+-/
+
+universe v u v₁ u₁
+
+open CategoryTheory LinearMap Opposite
+
+attribute [local instance] ConcreteCategory.hasCoeToSort ConcreteCategory.instFunLike
+
+variable {C : Type u₁} [Category.{v₁} C]
+
+namespace PresheafOfModules
+
+variable {R : Cᵒᵖ ⥤ CommRingCat.{u}}
+  (M : PresheafOfModules.{u} (R ⋙ forget₂ CommRingCat RingCat))
+
+/-- The type of absolute derivation from `R` to `M`, when `M` is a presheaf of modules
+over the presheaf of commutative rings `R`. -/
+@[ext]
+structure AbsoluteDerivation where
```

Comment:
I think we should define general derivations instead.

Reply:
A possible approach would be:
* define relative derivations and the universal property in the very general case `f : S ⟶ F.op ⋙ R`
* construct the $\Omega^1$ in the particular case `F` is the identity functor, i.e. just `f : S ⟶ R`
* using the pullback of presheaves of comm. rings, show that in general, derivations for `f : S ⟶ F.op ⋙ R` identify to derivations for $F^*(S) ⟶ R$, and then $\Omega^1$ also exists in that case.

(In the API is developed only for `f : S ⟶ R`, definitions may become ackward when we introduce $F$ because of the necessity to use $F^*$ despite the fact that the vanishing condition of relative derivations on the "base (presheaf of) rings" can be phrased in a cleaner manner.)

Reply:
I was thinking about https://stacks.math.columbia.edu/tag/01UN. 
I don't think I really understand the concern. If we define the structure `Derivation (f : S ⟶ R) (M)`, we can still provide specialized API for `Derivation (f : S ⟶ F.op ⋙ R) (M)` later on?

Reply:
> If we define the structure `Derivation (f : S ⟶ R) (M)`, we can still provide specialized API for `Derivation (f : S ⟶ F.op ⋙ R) (M)` later on?

No, because `M` should be a module over `R`, not over `F.op ⋙ R`.

My point is that if we only do https://stacks.math.columbia.edu/tag/01UN (which the "2." in my list above), then we would be forced to state things in terms the adjoint map `f^* O_S -> O_X` instead of the morphism `O_S -> f_* O_X` (which has much better definitional properties).

Reply:
To be more specific, if we want to pass from presheaves to sheaves, then the morphism `O_S -> f_* O_X` is the exact same whether we consider it as a morphism of sheaves or presheaves, while `f^* O_S` (or more precisely `f^{-1}`) is not the same object whether we take the inverse image as a sheaf or as a presheaf.

Reply:
This is restarted as #14014